HpToolsLauncher/ProcessExtensions.cs

```diff
@@ -44,5 +217,148 @@ public static Process Parent(this Process process)
+                {
+                    var si = (WTS_SESSION_INFO)Marshal.PtrToStructure((IntPtr)current, typeof(WTS_SESSION_INFO));
+                    current += arrayElementSize;
+
+                    if (si.State == WTS_CONNECTSTATE_CLASS.WTSActive)
+                    {
+                        activeSessionId = si.SessionID;
+                    }
+                }
+            }
+
+            // If enumerating did not work, fall back to the old method
+            if (activeSessionId == INVALID_SESSION_ID)
+            {
+                activeSessionId = WTSGetActiveConsoleSessionId();
+            }
+
+            if (WTSQueryUserToken(activeSessionId, ref hImpersonationToken) != 0)
+            {
+                // Convert the impersonation token to a primary token
+                bResult = DuplicateTokenEx(hImpersonationToken, 0, IntPtr.Zero,
+                    (int)SECURITY_IMPERSONATION_LEVEL.SecurityImpersonation, (int)TOKEN_TYPE.TokenPrimary,
+                    ref phUserToken);
+
+                CloseHandle(hImpersonationToken);
+            }
+
+            return bResult;
+        }
+
+        /// <summary>
+        /// Start a process from active user session
+        /// </summary>
+        /// <param name="appPath">The path of the module to be executed.</param>
+        /// <param name="cmdLine">The command line to be executed.</param>
+        /// <param name="workDir">The full path to the current directory for the process.</param>
+        /// <param name="visible">Show window or not.</param>
+        /// <returns>boolean result</returns>
+        public static int StartProcessFromUserSession(string appPath, string cmdLine = null, string workDir = null, bool visible = true)
+        {
+            var hUserToken = IntPtr.Zero;
+            var startInfo = new STARTUPINFO();
+            var procInfo = new PROCESS_INFORMATION();
+            var pEnv = IntPtr.Zero;
+            int iResultOfCreateProcessAsUser;
+            uint iExitCode = 0;
+
+            startInfo.cb = Marshal.SizeOf(typeof(STARTUPINFO));
+
+            try
+            {
+                if (!GetSessionUserToken(ref hUserToken))
+                {
+                    throw new Exception("StartProcessAsCurrentUser: GetSessionUserToken failed.");
+                }
+
+                uint dwCreationFlags = CREATE_UNICODE_ENVIRONMENT | (uint)(visible ? CREATE_NEW_CONSOLE : CREATE_NO_WINDOW);
+                startInfo.wShowWindow = (short)(visible ? SW.SW_SHOW : SW.SW_HIDE);
+                startInfo.lpDesktop = "winsta0\\default";
+                startInfo.hStdOutput = GetStdHandle(STD_OUTPUT_HANDLE);
+                startInfo.hStdError = GetStdHandle(STD_ERROR_HANDLE);
+                startInfo.dwFlags |= STARTF_USESTDHANDLES;
+                if (!CreateEnvironmentBlock(ref pEnv, hUserToken, false))
+                {
+                    throw new Exception("StartProcessAsCurrentUser: CreateEnvironmentBlock failed.");
```

Comment:
Do not throw exception here, it will terminate the program. Instead, return a flag to the caller, and if the flag is true, the caller shall ignore the **second** instance and continue all the processes in the current program in legacy behavior, that is, run test in the current **first** program.

Write output log (use `Console.WriteLine`) if the failed to create environment block.

It may not possible use the return value for this purpose, because the return value is the error code returned by the **second** program, it is impossible to distinguish between the case of "no user session" and the exit code of the **second** program.